File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/AdInstanceManager.kt
package com.cleveradssolutions.plugin.flutter

import android.app.Activity
import android.util.Size
import com.cleveradssolutions.sdk.AdContentInfo
import com.cleversolutions.ads.AdError
import io.flutter.plugin.common.MethodChannel

/**
 * Maintains reference to ad instances for the CASMobileAdsPlugin.
 *
 * When an Ad is loaded from Dart, an equivalent ad object is created
 * and maintained here to provide access until the ad is disposed.
 */
internal class AdInstanceManager(
    private val channel: MethodChannel
) {
    var activity: Activity? = null

    private val ads = HashMap<Int, FlutterAd>()

    fun findAd(adId: Int): FlutterAd? = ads[adId]

    fun trackAd(ad: FlutterAd) {
        require(ads[ad.adId] == null) {
            "Ad for following adId already exists: ${ad.adId}"
        }
        ads[ad.adId] = ad
    }

    fun disposeAd(adId: Int) {
        ads.remove(adId)?.dispose()
    }

    fun disposeAllAds() {
        for ((_, value) in ads) {
            value.dispose()
        }
        ads.clear()
    }

    fun onAdLoaded(adId: Int, size: Size?) {
        val args = HashMap<String, Any>()
        args["adId"] = adId
        if (size != null)
            args["size"] = size
        channel.invokeMethod("onAdLoaded", args)
    }

    fun onAdFailedToLoad(adId: Int, error: AdError) {
        val args = HashMap<String, Any>()
        args["adId"] = adId
        args["error"] = error
        channel.invokeMethod("onAdFailedToLoad", args)
    }

    fun onAdFailedToShow(adId: Int, error: AdError) {
        val args = HashMap<String, Any>()
        args["adId"] = adId
        args["error"] = error
        channel.invokeMethod("onAdFailedToShow", args)
    }

    fun onAdShowed(adId: Int) {
        invokeOnAdEvent("onAdShowed", adId)
    }

    fun onAdClicked(adId: Int) {
        invokeOnAdEvent("onAdClicked", adId)
    }

    fun onAdImpression(adId: Int, contentInfo: AdContentInfo) {
        val args = HashMap<String, Any>()
        args["adId"] = adId
        args["info"] = contentInfo
        channel.invokeMethod("onAdImpression", args)
    }

    fun onAdDismissed(adId: Int) {
        invokeOnAdEvent("onAdDismissed", adId)
    }

    fun onAdUserEarnedReward(adId: Int) {
        invokeOnAdEvent("onAdUserEarnedReward", adId)
    }

    private fun invokeOnAdEvent(name: String, adId: Int) {
        val args = HashMap<String, Any>()
        args["adId"] = adId
        channel.invokeMethod(name, args)
    }

}

File: ios/Classes/AdInstanceManager.swift
import CleverAdsSolutions
import Flutter

class AdInstanceManager {
    let channel: FlutterMethodChannel

    private var ads = [Int: FlutterAd]()
    private let adsLockQueue = DispatchQueue(label: "CASMobileAdsInstancesLockQueue", attributes: .concurrent)

    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }

    func findAd(_ adId: Int) -> FlutterAd? {
        adsLockQueue.sync {
            ads[adId]
        }
    }

    func trackAd(_ ad: FlutterAd) {
        adsLockQueue.async(flags: .barrier) {
            self.ads[ad.adId] = ad
        }
    }

    func disposeAd(_ adId: Int) {
        var ad: FlutterAd?
        adsLockQueue.sync {
            ad = self.ads.removeValue(forKey: adId)
        }
        ad?.disposeAd()
    }

    func disposeAllAds() {
        let all = ads
        ads = [Int: FlutterAd]()

        for item in all {
            item.value.disposeAd()
        }
    }

    func onAdLoaded(adId: Int, size: CGSize? = nil) {
        let args: [String: Any?] = [
            "adId": adId,
            "size": size,
        ]
        channel.invokeMethod("onAdLoaded", arguments: args)
    }

    func onAdFailedToLoad(adId: Int, error: AdError) {
        let args: [String: Any] = [
            "adId": adId,
            "error": error,
        ]
        channel.invokeMethod("onAdFailedToLoad", arguments: args)
    }

    func onAdFailedToShow(adId: Int, error: AdError) {
        let args: [String: Any] = [
            "adId": adId,
            "error": error,
        ]
        channel.invokeMethod("onAdFailedToShow", arguments: args)
    }

    func onAdShowed(adId: Int) {
        invokeOnAdEvent("onAdShowed", adId: adId)
    }

    func onAdClicked(adId: Int) {
        invokeOnAdEvent("onAdClicked", adId: adId)
    }

    func onAdImpression(adId: Int, contentInfo: AdContentInfo) {
        let args: [String: Any] = [
            "adId": adId,
            "info": contentInfo,
        ]
        channel.invokeMethod("onAdImpression", arguments: args)
    }

    func onAdDismissed(adId: Int) {
        invokeOnAdEvent("onAdDismissed", adId: adId)
    }

    func onAdUserEarnedReward(adId: Int) {
        invokeOnAdEvent("onAdUserEarnedReward", adId: adId)
    }

    private func invokeOnAdEvent(_ name: String, adId: Int) {
        channel.invokeMethod(name, arguments: ["adId": adId])
    }
}
